_posts/2023-04-30-writing-code-with-chatgpt-improve-it-with-kani.md

```diff
@@ -0,0 +1,369 @@
+
+Giving the above prompt, ChatGPT answered us with the following function
+
+```rust
+fn integer_average(a: i32, b: i32) -> i32 {
+    (a + b) / 2
+}
+```
+
+, an explanation
+>In this function, we define a function called integer_average that takes two integer arguments a and b.
+The function then adds the two arguments together and divides the result by 2 to compute the integer average.
+
+, and even a driver function to call and test the average function.
+
+```rust
+fn main() {
+    let a = 4;
+    let b = 8;
+    let result = integer_average(a, b);
+    println!("The average of {} and {} is {}", a, b, result);
+}
+```
+
+From the example, ChatGPT's code generation capabilities are thoroughly impressive.
+Not only does it produce code that looks good at first glance, but it also provides explanations that give us greater confidence that the code is doing exactly what we intended it to do.
+In fact, we can even request ChatGPT to generate a driver function and some test cases using follow-up prompts to further test the generated code.
+However, as verification folks, we know that we still can't be completely sure that the code is correct without **verification**.
+
+
+## Verification of Generated Code
+So, naturally, our next task is: verifying the generated programs with Kani.
+That is, we want to run Kani on the generated function `integer_average` to see if there is any bug in it.
+
+We can simply transform the driver function into a harness by substituting the test cases with `kani::any()`.
+
+``` rust
+fn integer_average(a: i32, b: i32) -> i32 {
+    (a + b) / 2     // <------ line 2
+}
+
+#[kani::proof]
+fn main() {
+    let a = kani::any();
+    let b = kani::any();
+    let result = integer_average(a, b);
+    println!("The average of {} and {} is {}", a, b, result);
+}
+```
+
+This harness verifies that, for any `i32` integers `a` and `b`, the function `integer_average` cannot trigger any of [bad behaviors that Kani checks for](https://model-checking.github.io/kani/tutorial-kinds-of-failure.html).
+
+We ran Kani on the above program
+
+```console
+> kani integer_average.rs
+```
+
+and got the verification result
+
+```console
+...
+SUMMARY:
+ ** 1 of 4 failed
+Failed Checks: attempt to add with overflow
+ File: "...", line 2, in integer_average
+
+VERIFICATION:- FAILED
+Verification Time: 0.6991158s
+```
+
+The code is actually buggy! 
+Although the average of two `i32` integers is in the range of `i32`, the intermediate sum `a + b` may overflow.
+A counterexample is when `a` equals `i32::MAX` and `b` equals any positive integer, the sum of them will be greater than `i32::MAX`, and hence the overflow.
+
+
+The examples above illustrate that, although code generation is efficient and accessible, it's crucial to exercise caution and avoid blindly trusting the generated code, as it may contain errors.
+We believe there are two factors that contribute to this unreliability.
+Firstly, the training corpus for most LLMs is unverified code.
```

Comment:
Can we make this claim? Is there a reference we can cite?

Reply:
I added a reference to codeX paper where they use code on github to train. 
And most of the program on github are unverified (I am not sure if I can find a reference for this).
Does it sound good to you?